File: crates/cgn-cli/tests/fixtures/receiver_types.php
<?php

class Animal {
    protected string $name;

    public function __construct(string $name) {
        $this->name = $name;
    }
}

class Dog extends Animal {
    public function bark(): void {
        $this->greet();
    }

    public function call_parent(): void {
        parent::__construct("parent_call");
    }

    public function call_self_static(): void {
        self::helper();
        static::helper();
    }

    public static function helper(): void {}

    protected function greet(): void {}
}

class Cat {
    public function meow(): void {
        $this->speak();
    }

    protected function speak(): void {}
}

function standalone(): void {
    $cat = new Cat();
    $cat->meow();
}

File: crates/cgn-cli/tests/fixtures/receiver_types.rs
struct Dog {
    name: String,
}

struct Cat;

impl Dog {
    fn new() -> Self {
        Dog { name: String::new() }
    }

    fn bark(&self) {
        self.fetch();
    }

    fn fetch(&self) {}
}

impl Cat {
    fn meow(&self) {}
}

trait Animal {
    fn speak(&self);
}

impl Animal for Dog {
    fn speak(&self) {
        self.bark();
    }
}

fn use_param(d: &Dog) {
    d.bark();
}

fn use_let_typed() {
    let d: Dog = Dog::new();
    d.bark();
}

fn use_let_inferred() {
    let d = Dog::new();
    d.bark();
}

fn use_scoped_path() {
    Dog::new();
}
